File: crates/keystone/sdk-python/src/tpt_sdk/wire.py
"""Client-side codec for the hand-written Postgres wire protocol v3 that
``tpt-keystone/src/wire/codec.rs`` implements server-side. This is a direct
Python port of the Rust SDK's ``tpt-sdk/src/keystone/wire.rs`` — read that
file's module doc first if you're comparing the two.

Only the subset of the protocol this SDK needs is implemented: the startup
handshake, the simple query protocol, and the extended query subset
(Parse/Bind/Describe/Execute/Sync) needed for parameterized queries. All
formats are text (format code 0) — there is no binary-format support, no
SASL/password auth (Keystone's startup handshake auto-approves any
``user`` parameter), and no SSL negotiation.

This module is deliberately dependency-free (stdlib ``asyncio`` only) — no
``psycopg2``/``asyncpg``/other Postgres driver. Pulling in a real Postgres
driver here would defeat the entire point of this project: every wire
protocol implementation (server *and* client) is hand-written from scratch.
"""

from __future__ import annotations

import asyncio
import struct
from dataclasses import dataclass, field
from typing import Optional


class WireError(Exception):
    """Raised for malformed frames or a connection that closed unexpectedly."""


class ServerError(Exception):
    """Raised when the server sends an ErrorResponse. Carries the
    human-readable message extracted from the 'M' field of the error
    fields (see ``parse_error_fields`` below); other fields (SQLSTATE code,
    severity, etc.) are not surfaced — matching the Rust SDK's scope cut.
    """


@dataclass
class FieldDescription:
    name: str
    type_oid: int


@dataclass
class RowDescriptionMsg:
    fields: list[FieldDescription]


@dataclass
class DataRowMsg:
    cells: list[Optional[bytes]]


@dataclass
class CommandCompleteMsg:
    tag: str


@dataclass
class ErrorResponseMsg:
    message: str


@dataclass
class NoticeResponseMsg:
    message: str


@dataclass
class ReadyForQueryMsg:
    status: int


@dataclass
class ParameterStatusMsg:
    name: str
    value: str


@dataclass
class BackendKeyDataMsg:
    pid: int
    secret: int


@dataclass
class ParameterDescriptionMsg:
    type_oids: list[int]


class AuthenticationOkMsg:
    pass


class ParseCompleteMsg:
    pass


class BindCompleteMsg:
    pass


class CloseCompleteMsg:
    pass


class NoDataMsg:
    pass


class PortalSuspendedMsg:
    pass


class EmptyQueryResponseMsg:
    pass


@dataclass
class UnknownMsg:
    tag: int


BackendMessage = object  # union of the *Msg classes above


def _read_cstr(buf: bytes, offset: int) -> tuple[str, int]:
    end = buf.index(b"\x00", offset)
    return buf[offset:end].decode("utf-8", errors="replace"), end + 1


def _parse_error_fields(buf: bytes, offset: int) -> str:
    """Error/notice fields are a sequence of (u8 field_code, cstr value)
    pairs terminated by a nul byte; we only surface the 'M' (message) field,
    matching the Rust SDK.
    """
    message: Optional[str] = None
    while offset < len(buf):
        code = buf[offset]
        offset += 1
        if code == 0:
            break
        value, offset = _read_cstr(buf, offset)
        if code == ord("M"):
            message = value
    return message if message is not None else "unknown server error"


class Conn:
    """One TCP connection to a Keystone node, speaking the wire protocol
    directly over ``asyncio`` streams (no connection pooling, no retries —
    a thin transport, same scope as the Rust SDK's ``Conn``).
    """

    def __init__(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
        self._reader = reader
        self._writer = writer
        self._write_buf = bytearray()

    @classmethod
    async def connect(cls, host: str, port: int, params: dict[str, str]) -> "Conn":
        reader, writer = await asyncio.open_connection(host, port)
        conn = cls(reader, writer)
        conn._write_startup(params)
        await conn.flush()

        while True:
            msg = await conn.read_message()
            if isinstance(msg, AuthenticationOkMsg):
                continue
            if isinstance(msg, ReadyForQueryMsg):
                break
            if isinstance(msg, ErrorResponseMsg):
                writer.close()
                raise ServerError(f"startup rejected: {msg.message}")
        return conn

    # -- message writers -------------------------------------------------

    def _write_startup(self, params: dict[str, str]) -> None:
        body = bytearray()
        body += struct.pack("!i", 196608)  # protocol version 3.0
        for k, v in params.items():
            body += k.encode("utf-8") + b"\x00"
            body += v.encode("utf-8") + b"\x00"
        body += b"\x00"
        self._write_buf += struct.pack("!i", 4 + len(body))
        self._write_buf += body

    def write_query(self, sql: str) -> None:
        def body(b: bytearray) -> None:
            b += sql.encode("utf-8") + b"\x00"

        self._write_msg(ord("Q"), body)

    def write_parse(self, name: str, sql: str, param_types: list[int]) -> None:
        def body(b: bytearray) -> None:
            b += name.encode("utf-8") + b"\x00"
            b += sql.encode("utf-8") + b"\x00"
            b += struct.pack("!h", len(param_types))
            for ty in param_types:
                b += struct.pack("!i", ty)

        self._write_msg(ord("P"), body)

    def write_bind(self, portal: str, stmt: str, params: list[Optional[bytes]]) -> None:
        def body(b: bytearray) -> None:
            b += portal.encode("utf-8") + b"\x00"
            b += stmt.encode("utf-8") + b"\x00"
            b += struct.pack("!h", 1)
            b += struct.pack("!h", 0)  # all params text format
            b += struct.pack("!h", len(params))
            for p in params:
                if p is None:
                    b += struct.pack("!i", -1)
                else:
                    b += struct.pack("!i", len(p))
                    b += p
            b += struct.pack("!h", 1)
            b += struct.pack("!h", 0)  # all results text format

        self._write_msg(ord("B"), body)

    def write_describe_portal(self, name: str) -> None:
        def body(b: bytearray) -> None:
            b += b"P"
            b += name.encode("utf-8") + b"\x00"

        self._write_msg(ord("D"), body)

    def write_execute(self, portal: str, max_rows: int) -> None:
        def body(b: bytearray) -> None:
            b += portal.encode("utf-8") + b"\x00"
            b += struct.pack("!i", max_rows)

        self._write_msg(ord("E"), body)

    def write_sync(self) -> None:
        self._write_msg(ord("S"), lambda b: None)

    def write_terminate(self) -> None:
        self._write_msg(ord("X"), lambda b: None)

    def _write_msg(self, tag: int, body_fn) -> None:
        b = bytearray()
        body_fn(b)
        self._write_buf.append(tag)
        self._write_buf += struct.pack("!i", 4 + len(b))
        self._write_buf += b

    async def flush(self) -> None:
        self._writer.write(bytes(self._write_buf))
        await self._writer.drain()
        self._write_buf.clear()

    # -- message reader ----------------------------------------------------

    async def read_message(self) -> BackendMessage:
        header = await self._readexactly(5)
        tag = header[0]
        (length,) = struct.unpack("!i", header[1:5])
        if length < 4:
            raise WireError(f"invalid message length {length} for tag {chr(tag)!r}")
        body = await self._readexactly(length - 4)

        offset = 0

        if tag == ord("R"):
            return AuthenticationOkMsg()
        if tag == ord("S"):
            name, offset = _read_cstr(body, offset)
            value, offset = _read_cstr(body, offset)
            return ParameterStatusMsg(name, value)
        if tag == ord("K"):
            pid, secret = struct.unpack("!ii", body[:8])
            return BackendKeyDataMsg(pid, secret)
        if tag == ord("Z"):
            status = body[0] if body else ord("I")
            return ReadyForQueryMsg(status)
        if tag == ord("T"):
            (n,) = struct.unpack("!h", body[offset : offset + 2])
            offset += 2
            fields = []
            for _ in range(n):
                name, offset = _read_cstr(body, offset)
                # table_oid(i32), col_attr(i16), type_oid(i32), type_size(i16),
                # type_modifier(i32), format(i16)
                _table_oid, _col_attr, type_oid, _type_size, _type_modifier, _format = struct.unpack(
                    "!ihihih", body[offset : offset + 18]
                )
                offset += 18
                fields.append(FieldDescription(name, type_oid))
            return RowDescriptionMsg(fields)
        if tag == ord("D"):
            (n,) = struct.unpack("!h", body[offset : offset + 2])
            offset += 2
            cells: list[Optional[bytes]] = []
            for _ in range(n):
                (clen,) = struct.unpack("!i", body[offset : offset + 4])
                offset += 4
                if clen < 0:
                    cells.append(None)
                else:
                    cells.append(body[offset : offset + clen])
                    offset += clen
            return DataRowMsg(cells)
        if tag == ord("C"):
            tag_str, offset = _read_cstr(body, offset)
            return CommandCompleteMsg(tag_str)
        if tag == ord("E"):
            return ErrorResponseMsg(_parse_error_fields(body, offset))
        if tag == ord("N"):
            return NoticeResponseMsg(_parse_error_fields(body, offset))
        if tag == ord("1"):
            return ParseCompleteMsg()
        if tag == ord("2"):
            return BindCompleteMsg()
        if tag == ord("3"):
            return CloseCompleteMsg()
        if tag == ord("t"):
            (n,) = struct.unpack("!h", body[offset : offset + 2])
            offset += 2
            types = []
            for _ in range(n):
                (ty,) = struct.unpack("!i", body[offset : offset + 4])
                offset += 4
                types.append(ty)
            return ParameterDescriptionMsg(types)
        if tag == ord("n"):
            return NoDataMsg()
        if tag == ord("s"):
            return PortalSuspendedMsg()
        if tag == ord("I"):
            return EmptyQueryResponseMsg()
        return UnknownMsg(tag)

    async def _readexactly(self, n: int) -> bytes:
        try:
            return await self._reader.readexactly(n)
        except asyncio.IncompleteReadError as exc:
            raise WireError("connection closed by server") from exc

    def close(self) -> None:
        self._writer.close()

    async def wait_closed(self) -> None:
        await self._writer.wait_closed()

File: sdk-go/wire.go
package keystone

// Hand-written client-side codec for the Postgres wire protocol v3 that
// tpt-keystone/src/wire implements server-side. This mirrors
// tpt-sdk/src/keystone/wire.rs (the Rust SDK's client codec) message-for-
// message: frontend (client -> server) messages are *encoded* here, backend
// (server -> client) messages are *decoded* here. Only the subset of the
// protocol this SDK needs is implemented: startup, the simple query loop,
// and the extended query subset (Parse/Bind/Describe/Execute/Sync) needed
// for parameterized and streaming queries. All formats are text (format
// code 0) — there is no binary-format support, matching the Rust SDK.
//
// No third-party Postgres driver (lib/pq, jackc/pgx, ...) is used anywhere
// in this module by design — see the project's CLAUDE.md.

import (
	"bufio"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net"
)

// fieldDescription describes one result column, as sent in a RowDescription
// message.
type fieldDescription struct {
	Name    string
	TypeOID int32
}

// msgKind tags which variant a decoded backendMessage holds.
type msgKind int

const (
	msgAuthenticationOk msgKind = iota
	msgParameterStatus
	msgBackendKeyData
	msgReadyForQuery
	msgRowDescription
	msgDataRow
	msgCommandComplete
	msgErrorResponse
	msgNoticeResponse
	msgParseComplete
	msgBindComplete
	msgCloseComplete
	msgParameterDescription
	msgNoData
	msgPortalSuspended
	msgEmptyQueryResponse
	msgUnknown
)

// backendMessage is a decoded server->client message. Only the fields
// relevant to msgKind are populated; this plays the role of the Rust SDK's
// `enum BackendMessage`.
type backendMessage struct {
	kind kindAndPayload
}

// kindAndPayload avoids needing one struct field per variant scattered
// across call sites; fields are still named per-variant for clarity.
type kindAndPayload struct {
	Kind        msgKind
	ParamName   string
	ParamValue  string
	PID         int32
	Secret      int32
	ReadyStatus byte
	Fields      []fieldDescription
	Row         []*[]byte // nil element == SQL NULL
	CommandTag  string
	ErrMsg      string
	ParamTypes  []int32
	UnknownTag  byte
}

var errConnClosed = errors.New("keystone: connection closed by server")

// wireConn is the low-level framer: it knows how to write frontend messages
// into a buffer and decode backend messages from a bufio.Reader. It does
// not know about contexts; deadline/cancellation plumbing lives in Conn
// (conn.go), which wraps wireConn.
type wireConn struct {
	nc      net.Conn
	r       *bufio.Reader
	wbuf    []byte
	hdrbuf  [5]byte
	lenbuf  [4]byte
	readBuf []byte // reused scratch buffer for message bodies
}

func newWireConn(nc net.Conn) *wireConn {
	return &wireConn{nc: nc, r: bufio.NewReaderSize(nc, 16*1024)}
}

// --- frontend message encoding -------------------------------------------------

func (w *wireConn) resetWrite() {
	w.wbuf = w.wbuf[:0]
}

func putI32(buf []byte, v int32) []byte {
	var b [4]byte
	binary.BigEndian.PutUint32(b[:], uint32(v))
	return append(buf, b[:]...)
}

func putI16(buf []byte, v int16) []byte {
	var b [2]byte
	binary.BigEndian.PutUint16(b[:], uint16(v))
	return append(buf, b[:]...)
}

func putCStr(buf []byte, s string) []byte {
	buf = append(buf, s...)
	buf = append(buf, 0)
	return buf
}

// writeStartup queues the StartupMessage (no auth type byte — it's the one
// message with no leading tag byte).
func (w *wireConn) writeStartup(params map[string]string) {
	var body []byte
	body = putI32(body, 196608) // protocol version 3.0
	for k, v := range params {
		body = putCStr(body, k)
		body = putCStr(body, v)
	}
	body = append(body, 0)
	w.wbuf = putI32(w.wbuf, int32(4+len(body)))
	w.wbuf = append(w.wbuf, body...)
}

func (w *wireConn) writeMsg(tag byte, body []byte) {
	w.wbuf = append(w.wbuf, tag)
	w.wbuf = putI32(w.wbuf, int32(4+len(body)))
	w.wbuf = append(w.wbuf, body...)
}

func (w *wireConn) writeQuery(sql string) {
	w.writeMsg('Q', putCStr(nil, sql))
}

func (w *wireConn) writeParse(name, sql string, paramTypes []int32) {
	var b []byte
	b = putCStr(b, name)
	b = putCStr(b, sql)
	b = putI16(b, int16(len(paramTypes)))
	for _, t := range paramTypes {
		b = putI32(b, t)
	}
	w.writeMsg('P', b)
}

func (w *wireConn) writeBind(portal, stmt string, params []*[]byte) {
	var b []byte
	b = putCStr(b, portal)
	b = putCStr(b, stmt)
	b = putI16(b, 1)
	b = putI16(b, 0) // all params text format
	b = putI16(b, int16(len(params)))
	for _, p := range params {
		if p == nil {
			b = putI32(b, -1)
			continue
		}
		b = putI32(b, int32(len(*p)))
		b = append(b, (*p)...)
	}
	b = putI16(b, 1)
	b = putI16(b, 0) // all results text format
	w.writeMsg('B', b)
}

func (w *wireConn) writeDescribePortal(name string) {
	b := []byte{'P'}
	b = putCStr(b, name)
	w.writeMsg('D', b)
}

func (w *wireConn) writeExecute(portal string, maxRows int32) {
	b := putCStr(nil, portal)
	b = putI32(b, maxRows)
	w.writeMsg('E', b)
}

func (w *wireConn) writeSync() {
	w.writeMsg('S', nil)
}

func (w *wireConn) writeTerminate() {
	w.writeMsg('X', nil)
}

// flush blocks writing the queued bytes; callers apply ctx deadlines before
// calling this (see conn.go).
func (w *wireConn) flush() error {
	if len(w.wbuf) == 0 {
		return nil
	}
	_, err := w.nc.Write(w.wbuf)
	w.resetWrite()
	return err
}

// --- backend message decoding -------------------------------------------------

func readCStr(data []byte, pos *int) string {
	start := *pos
	end := start
	for end < len(data) && data[end] != 0 {
		end++
	}
	s := string(data[start:end])
	if end < len(data) {
		end++ // skip the nul
	}
	*pos = end
	return s
}

func parseErrorFields(data []byte) string {
	pos := 0
	message := ""
	for pos < len(data) {
		code := data[pos]
		pos++
		if code == 0 {
			break
		}
		val := readCStr(data, &pos)
		if code == 'M' {
			message = val
		}
	}
	if message == "" {
		message = "unknown server error"
	}
	return message
}

// readMessage blocks reading exactly one backend message; callers apply ctx
// deadlines/cancellation before calling this (see conn.go's withDeadline).
func (w *wireConn) readMessage() (kindAndPayload, error) {
	if _, err := readFull(w.r, w.hdrbuf[:5]); err != nil {
		return kindAndPayload{}, err
	}
	tag := w.hdrbuf[0]
	length := int32(binary.BigEndian.Uint32(w.hdrbuf[1:5]))
	if length < 4 {
		return kindAndPayload{}, fmt.Errorf("keystone: invalid message length %d for tag %q", length, tag)
	}
	bodyLen := int(length) - 4
	if cap(w.readBuf) < bodyLen {
		w.readBuf = make([]byte, bodyLen)
	}
	body := w.readBuf[:bodyLen]
	if bodyLen > 0 {
		if _, err := readFull(w.r, body); err != nil {
			return kindAndPayload{}, err
		}
	}
	return decodeBody(tag, body)
}

func readFull(r *bufio.Reader, buf []byte) (int, error) {
	n := 0
	for n < len(buf) {
		m, err := r.Read(buf[n:])
		n += m
		if err != nil {
			if n == len(buf) {
				break
			}
			if errors.Is(err, io.EOF) && n == 0 {
				return n, errConnClosed
			}
			return n, err
		}
	}
	return n, nil
}

func decodeBody(tag byte, data []byte) (kindAndPayload, error) {
	pos := 0
	switch tag {
	case 'R':
		return kindAndPayload{Kind: msgAuthenticationOk}, nil
	case 'S':
		name := readCStr(data, &pos)
		value := readCStr(data, &pos)
		return kindAndPayload{Kind: msgParameterStatus, ParamName: name, ParamValue: value}, nil
	case 'K':
		pid := int32(binary.BigEndian.Uint32(data[0:4]))
		secret := int32(binary.BigEndian.Uint32(data[4:8]))
		return kindAndPayload{Kind: msgBackendKeyData, PID: pid, Secret: secret}, nil
	case 'Z':
		status := byte('I')
		if len(data) > 0 {
			status = data[0]
		}
		return kindAndPayload{Kind: msgReadyForQuery, ReadyStatus: status}, nil
	case 'T':
		n := int(binary.BigEndian.Uint16(data[0:2]))
		pos = 2
		fields := make([]fieldDescription, 0, n)
		for i := 0; i < n; i++ {
			name := readCStr(data, &pos)
			pos += 4 // table oid
			pos += 2 // col attr
			typeOID := int32(binary.BigEndian.Uint32(data[pos : pos+4]))
			pos += 4
			pos += 2 // type size
			pos += 4 // type modifier
			pos += 2 // format code
			fields = append(fields, fieldDescription{Name: name, TypeOID: typeOID})
		}
		return kindAndPayload{Kind: msgRowDescription, Fields: fields}, nil
	case 'D':
		n := int(binary.BigEndian.Uint16(data[0:2]))
		pos = 2
		cols := make([]*[]byte, 0, n)
		for i := 0; i < n; i++ {
			l := int32(binary.BigEndian.Uint32(data[pos : pos+4]))
			pos += 4
			if l < 0 {
				cols = append(cols, nil)
				continue
			}
			take := int(l)
			if pos+take > len(data) {
				take = len(data) - pos
			}
			cell := make([]byte, take)
			copy(cell, data[pos:pos+take])
			pos += take
			cols = append(cols, &cell)
		}
		return kindAndPayload{Kind: msgDataRow, Row: cols}, nil
	case 'C':
		return kindAndPayload{Kind: msgCommandComplete, CommandTag: readCStr(data, &pos)}, nil
	case 'E':
		return kindAndPayload{Kind: msgErrorResponse, ErrMsg: parseErrorFields(data)}, nil
	case 'N':
		return kindAndPayload{Kind: msgNoticeResponse, ErrMsg: parseErrorFields(data)}, nil
	case '1':
		return kindAndPayload{Kind: msgParseComplete}, nil
	case '2':
		return kindAndPayload{Kind: msgBindComplete}, nil
	case '3':
		return kindAndPayload{Kind: msgCloseComplete}, nil
	case 't':
		n := int(binary.BigEndian.Uint16(data[0:2]))
		pos = 2
		types := make([]int32, 0, n)
		for i := 0; i < n; i++ {
			types = append(types, int32(binary.BigEndian.Uint32(data[pos:pos+4])))
			pos += 4
		}
		return kindAndPayload{Kind: msgParameterDescription, ParamTypes: types}, nil
	case 'n':
		return kindAndPayload{Kind: msgNoData}, nil
	case 's':
		return kindAndPayload{Kind: msgPortalSuspended}, nil
	case 'I':
		return kindAndPayload{Kind: msgEmptyQueryResponse}, nil
	default:
		return kindAndPayload{Kind: msgUnknown, UnknownTag: tag}, nil
	}
}
